### DISCUSSION TITLE
RFC: A GeoParquet Index File to Improve Spatial Query Performance

### DISCUSSION THREAD

**Discussion by jwass** [OvertureMaps/data#219] [Discussion]
## Background

Overture releases data as a hive-partitioned GeoParquet dataset including the bounding box (bbox) covering column. Our Spark/Sedona pipeline creates file sizes of ~1GB which can result in tens or even hundreds of GeoParquet files within a single partition.

To perform a spatial query on a GeoParquet dataset, clients must first issue several network requests for each Parquet file to acquire the Parquet metadata. Using the metadata's row group summary statistics on the bbox covering column, clients next download the row group data and finally filter for the rows overlapping a bounding box region of interest.

In the end, clients may make hundreds or even thousands of network requests and download tens of megabytes of parquet metadata before acquiring the desired rows. Although this process is orders of magnitude faster than downloading the entire GeoParquet dataset it can still be improved.

We propose creating an index file that reduces the number of files that a client must consult when querying for a region of interest.

## Index File Proposal

The index file is a single GeoParquet file that contains the filename, total bounding box of geometries in that file, and a geometry representing the file's extent.

It has the following columns:
- filename: Varchar - the relative path of the Parquet file
- bbox: a GeoParquet bbox (https://github.com/opengeospatial/geoparquet/blob/main/format-specs/geoparquet.md#bbox)
- geometry: Varbinary - A WKB polygon with four points that are the corners of the bbox
- additional columns are allowed.

Clients can download the small index file, filter for the rows that overlap a bounding box. The resulting rows contain the Parquet filenames containing the row-level data. Client next acquire the Parquet metadata and row groups from those files. 

The existence of an index file is backwards-compatible in that clients that do not have capabilities of reading an index file will use the same GeoParquet access over the whole dataset without any regressions.

### Format
This proposal chooses GeoParquet as the spatial filtering logic for the rows of the index file is identical when used in the data files.

There was some discussion about using a JSON file for the index file. We decided to use GeoParquet because clients will need a GeoParquet reader for the data files anyway.

### Where Do Index Files Go? In Their Own index/ prefix
Example:
Update: (Modified after conversation about Overture metadata)
```
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=buildings/type=building/index.parquet
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=buildings/type=part/index.parquet
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=addresses/type=address/index.parquet
```

### Versioning (TBD)
If we change the schema or other behaviors of an index file, how will clients know which kind they are using? We could include a version field in the Parquet header similar how GeoParquet includes its top-level key-value metadata.

If we change the index file, would we recompute it for older Overture versions?

## Prior Art
- We've had many discussions on this topic
- @Bonkles has experimented with a JSON summary file (https://github.com/Bonkles/manifest-generator)
- @RaczeQ proposed a similar version but can break down file-level bounding boxes further (https://www.linkedin.com/posts/raczyckikamil_python-pyarrow-parquet-activity-7231632304907931648-_zc_?utm_source=share&utm_medium=member_desktop).

## Implementation
Update:
Here's an example of generating an implementation and a script to compare using vs. not using the index - for both geoarrow and duckdb. https://github.com/jwass/gpqindex

**Comment by RaczeQ** [comment]
Here is the repository with precalculated indexes based on my method: https://github.com/kraina-ai/overturemaps-releases-indexes
In the end, I've decided to consider @jwass suggestion regarding keeping bounding boxes separate and I've added a column with list of IDs within a given row group (in a form of compressed ranges).

This step with having multiple bounding boxes per single row group is probably too complex to have it in an official index, but having a bounding box per file would improve performance if utilised by the query engine.

**Reply by jwass** [reply]
The main thing I'm targeting here is the network traffic - both number of connections and data transferred. If a Parquet client knows which row group to read, it still needs to get all the metadata of that Parquet file anyway- unless we do something very different. So the index helping narrow down to specific row groups doesn't help that much and just narrowing down the files seems like the biggest impact.

What I like about the approach you posted on LinkedIn is that it can break down the large bounding box of a single file into several smaller ones which should help with further pruning out files to query. But the index is more intense to create since you need to process and iterate over all the geometries, but probably has some value.

**Comment by mtravis** [comment]
@jwass This sounds fantastic. Anything that speeds up bbox querying gets my vote! 

Happy to help with any testing with DuckDB, pyarrow, etc.

**Reply by vcschapp** [reply]
Would tooling like DuckDB be able to understand the index file, or is this more directed towards custom solutions like data pipelines?

**Reply by RaczeQ** [reply]
It would be possible to write a custom query for loading intersected rows, get filenames and pass them as an argument to the `read_parquet` function.

Here is an example: https://github.com/duckdb/duckdb/discussions/8144

**Reply by mtravis** [reply]
@vcschapp I'm assuming that all clients that can read geoparquet will be able make use of this and this is an improvement for end users as well as internal pipelines?

**Reply by jwass** [reply]
@vcschapp See https://github.com/jwass/gpqindex/blob/main/runner.py#L67 for how we can use duckdb to make use of the index file.

**Reply by jwass** [reply]
@RaczeQ I don't believe we can make this work in one query within duckdb. The `read_parquet()` call needs to have constant inputs that can't be the result of a subquery, as I understand it anyway. But doing it in two queries doesn't seem particularly bad.

**Comment by jwass** [comment]
Sample code for both generating an index like this - and code for using it (in both pyarrow and duckdb) is here: https://github.com/jwass/gpqindex/. For buildings only.

I put a sample _index.parquet file generated by the script in a non-public bucket but anyone can generate it and try this out themselves. 

You can run:
```
$ python runner.py --pyarrow-index
$ python runner.py --pyarrow-no-index
$ python runner.py --duckdb-index
$ python runner.py --duckdb-no-index
```

to compare different times for pyarrow or duckdb both with and without using the index. Seeing at least 5x improvements in runtime when using it. If anyone else wants to give it a spin, would love to see your results.

**Reply by mtravis** [reply]
@jwass 

Trying to test locally and hitting an issue when trying to create the `_index.parquet`

Seeing this error: `KeyError: "['row_group_index', 'num_rows'] not found in axis"
` What's the reason for dropping these columns?

I'm liking the sound of 5x speed improvements by the way!

**Reply by jwass** [reply]
@mtravis Can you pull the latest?

**Reply by mtravis** [reply]
@jwass I'm up-to-date. Have you made some new commits?

**Reply by jwass** [reply]
Ooops. Hadn't actually pushed that fix. Try again

**Reply by mtravis** [reply]
```
python runner.py --duckdb-index
Run 0: 6.606157064437866 s
Run 1: 4.932492017745972 s
Run 2: 4.506331443786621 s
Run 3: 4.704394817352295 s
Run 4: 4.419393301010132 s
Median time is 4.704394817352295

python3  runner.py --duckdb-no-index
Run 0: 296.20549750328064 s
Run 1: 286.11207342147827 s
Run 2: 298.4442949295044 s
Run 3: 295.81089782714844 s
Run 4: 291.1927001476288 s
Median time is 295.81089782714844
```

**Comment by jiayuasu** [comment]
@jwass In the Iceberg Geometry proposal, we are proposing to have a partition transform mechanism for the geometry column. With that, the spatial query will jump to the partition directly without going the parquet metadata.

Partition transform is a unique and powerful feature of Iceberg.

**Reply by jwass** [reply]
@jiayuasu Thanks. That sounds great. I think it will be a while before Overture switches away from GeoParquet so I'm interested in seeing what we can push with these kinds of improvements.

Looking forward to seeing the Iceberg natively support geometry!

**Comment by barbuz** [comment]
This sounds like a great idea! I have a couple of questions:

- I am not sure about where are you proposing to store the index files. If they are in the same folder as the data files how can this be backwards-compatible? If they are in a different folder, how will a reader find them?
- Could this proposal be extensible to act as an index on arbitrary columns? I'm thinking of other use cases where a dataset might be split into files based on some variable other than space (e.g. time?), then the index file could store the min-max values of arbitrary columns in the data to generate an index over those columns.

**Reply by jwass** [reply]
Hi @barbuz - thanks for the questions.

> I am not sure about where are you proposing to store the index files. If they are in the same folder as the data files how can this be backwards-compatible? If they are in a different folder, how will a reader find them?

Yeah I don't think there's a way to store them alongside the main parquet data files. We're instead looking at a top-level `bbox_index` (exact name TBD) prefix that will mimic the partitioning structure. Something like this:

```
s3://overturemaps-us-west-2/release/2024-08-20.0/theme=buildings/type=building/.....parquet <-- the existing file structure of the data files
 
bbox index files:
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=buildings/type=building/index.parquet
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=base/type=infrastructure/index.parquet
s3://overturemaps-us-west-2/bbox_index/2024-08-20.0/theme=base/type=water/index.parquet
...
```
Not idea, but maintains our current file structure

> Could this proposal be extensible to act as an index on arbitrary columns? I'm thinking of other use cases where a dataset might be split into files based on some variable other than space (e.g. time?), then the index file could store the min-max values of arbitrary columns in the data to generate an index over those columns.

I think so. One issue this might have is that the main data files have a rough spatial ordering so the bbox index technique works well. But other columns won't necessarily be sorted that way so this min-max technique might have diminishing returns. There are other things to look at like individual row-level indexing or something similar to help narrow down the search space.

**Comment by yharby** [comment]
What about https://github.com/stac-utils/stac-geoparquet ? … Can it be used as the index file with adopting STAC ?

**Reply by mtravis** [reply]
@Youssef-Harby Nice idea but I believe that `stac-geoparquet` is specifically for use with STAC end points only.

**Comment by pradeepbkkumar** [comment]
@jwass I haven't worked on Parquet/GeoParquet files much.  I am just getting started. So, how long does it take for the clients to read the row-groups from 1GB parquet file for both spatial and non-spatial queries? Is 1GB the right partition size? Your index file size will grow/shrink depending on the number of partitions. At MSFT, we have a legacy store where we use text files (we haven't moved to parquet yet), so we keep the partition size too small. Each partition will have only 3000 entities (features)
. It leads to ~1M partitions and because of that, the index file size is huge and so we create a range index file that further indexes the index file
